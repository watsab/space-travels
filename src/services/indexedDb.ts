import type { Pilot } from '@/models/Pilot';

let database: IDBDatabase;
export const getDatabase = (): Promise<IDBDatabase> => {
  return new Promise((resolve, reject) => {
    if (database) {
      return resolve(database)
    }

    let request: IDBOpenDBRequest = window.indexedDB.open('space-travels', 1)

    request.onerror = (event: Event) => {
      console.error('ERROR: Unable to open database', event);
      reject('Error');
    }

    request.onsuccess = (event: Event) => {
      database = event.target.result;
      resolve(database);
    }

    // Hook appelé à la création ou mise à jour de la bdd
    request.onupgradeneeded = (event: Event) => {
      let database: IDBDatabase = event.target.result;
      database.createObjectStore('pilots', {
        autoIncrement: true,
        keyPath: 'id',
      });
    }
  })
};


export const getPilots = (database: IDBDatabase): Promise<Pilot[]> => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readonly');
    const store = transaction.objectStore('pilots');
    const pilots: Pilot[] = [];

    store.openCursor().onsuccess = (event: Event) => {
      let cursor = event.target.result;
      if (cursor) {
        pilots.push(cursor.value as Pilot);
        cursor.continue();
      }
    }

    transaction.oncomplete = () => {
      resolve(pilots)
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};
export const getPilot = (database: IDBDatabase, pilotId: number): Promise<Pilot[]> => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readonly');
    const store = transaction.objectStore('pilots');

    let pilot: Pilot;
    store.openCursor().onsuccess = (event: Event) => {
      let cursor = event.target.result;
      if (cursor) {
        if (cursor.value.id === pilotId) {
          pilot = cursor.value;
        }
        cursor.continue();
      }
    }

    transaction.oncomplete = () => {
      resolve(pilot)
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};


export const addPilot = (database: IDBDatabase, pilot: Pilot): Promise<boolean> => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readwrite');
    const store = transaction.objectStore('pilots');

    store.put({...pilot});

    transaction.oncomplete = () => {
      console.log('Pilot successfully saved.');
      resolve(true)
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};

export const removePilot = (database: IDBDatabase, pilot: Pilot) => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readwrite');
    const store = transaction.objectStore('pilots');

    store.delete(pilot.id);
    transaction.oncomplete = () => {
      console.log('Pilot successfully saved.');
      resolve(true);
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};
