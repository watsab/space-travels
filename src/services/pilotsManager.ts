import type { Pilot } from '@/models/Pilot';
import pilotData from '@/data/pilots.json';

const getAll = (): Pilot[] => {
  return pilotData.map((pilot: Pilot): Pilot => {
    return {
      ...pilot,
      picturePath: new URL(`../assets/pilots/${pilot.picturePath}`, import.meta.url).toString()
    }
  });
}

const getById = (pilotId: number): Pilot | undefined => getAll().find(({ id }) => id === pilotId);

export {
  getAll,
  getById
}
