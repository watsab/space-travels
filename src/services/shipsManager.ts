import { usePilotsManager } from '@/composable/usePilotsManager';
import type { Ship } from '@/models/Ship';
import shipData from '@/data/ships.json';


const getAll = (): Ship[] => {
  const pilots = usePilotsManager().getAll();
  return shipData.map((ship): Ship => {
    return {
      name: ship.name,
      slug: ship.slug,
      pilot: pilots.find(({ id }) => ship.pilotId === id)
    }
  });
}

const getBySlug = (shipSlug: string): Ship | undefined => {
  getAll().find(({ slug }) => slug === shipSlug)
}

export {
  getAll,
  getBySlug,
}
