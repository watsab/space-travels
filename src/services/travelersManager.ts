import type { Traveler } from '@/models/Traveler';

const getAll = async (page = 1, limit = 5) => {
  const response = await fetch(`https://jsonplaceholder.typicode.com/users?_limit=${limit}&_page=${page}`);
  if (!response.ok) {
    throw new Error('An error occurred');
  }
  return await response.json() as Traveler[]
};

const getById = async (id: number): Promise<Traveler | null> => {
  const response = await fetch(`https://jsonplaceholder.typicode.com/users/${id}`);
  if (!response.ok) {
    if (response.status === 404) {
      return null;
    }

    throw new Error('An error occurred');
  }
  return await response.json() as Traveler;
}


export {
  getAll,
  getById,
}
