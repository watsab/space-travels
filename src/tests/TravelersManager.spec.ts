import { describe, it, expect, vi } from 'vitest';
import { Response } from 'happy-dom';
import { getById } from '@/services/travelersManager';
import type { Mock } from '@vitest/spy';

const mocks: { response: { body: undefined, init: {status: number}},  fetch: Mock} = vi.hoisted(() => ({
  response: {
    body: undefined,
    init: {
      status: 200
    }
  },
  fetch: vi.fn(() =>
    new Response(mocks.response.body, mocks.response.init)
  )}
));

vi.stubGlobal('fetch', mocks.fetch);

describe('TravelersManager', () => {
  it('The getId method throws an error when the status code is invalid and not a 404', async () => {
    mocks.response.init.status = 500;
    await expect(getById(123)).rejects.toThrowError('An error');
  });

  it.todo('The getId method returns null the status code is a 404', async () => {
    mocks.response.init.status = 404;
    await expect(getById(123)).resolves.toBeNull();

  });
});
