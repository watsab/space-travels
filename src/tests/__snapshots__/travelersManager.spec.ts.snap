// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Traveler/ListView > The rendering matches the snapshot 1`] = `
"<h2 data-v-90230f27=\\"\\">Liste des voyageurs</h2>
<my-data-table-stub data-v-90230f27=\\"\\" items=\\"\\"></my-data-table-stub>
<div data-v-90230f27=\\"\\" class=\\"pagination\\">
  <div data-v-90230f27=\\"\\">
    <router-link-stub data-v-90230f27=\\"\\" to=\\"[object Object]\\" rel=\\"prev\\" style=\\"display: none;\\"></router-link-stub>
  </div>
  <div data-v-90230f27=\\"\\">
    <router-link-stub data-v-90230f27=\\"\\" to=\\"[object Object]\\" rel=\\"next\\"></router-link-stub>
  </div>
</div>"
`;
